ENH: spatial.geometric_slerp: allow for extrapolation

tylerjereddy opened the pull request:
* Fixes gh-23400.

* Add support and regression tests for extrapolation behavior in `geometric_slerp()`, and update the documentation for the interpolation parameter accordingly.

* The `Raises` section of the `geometric_slerp()` docstring was also updated to remove a spurious claim about antipodes--we now actually allow i.e., North and South pole points despite the infinite number of paths between them on the sphere, and enforce this behavior in our regression tests (a warning, but not error, is issued). That is not directly related to the changes here, but the documented innacurracy of behavior would otherwise make this harder to reason about.

[ci skip] [skip ci]

TODO:

- [x] reactivate the CI and make sure it is passing (it passes locally, for now)
- [x] think of some more useful extrapolation test cases to add?
- [x] expand the `geometric_slerp()` docstring to include an example (+ plot) of extrapolation on the circle or sphere

@Ockenfuss since you were enthusiastic about adding this capability, would you like to suggest a few more interesting test cases and possibly make a suggestion for a docstring example of extrapolation you'd like to see?

tylerjereddy committed:
ENH: geometric_slerp extrapolation

* Fixes gh-23400.

* Add support and regression tests for extrapolation behavior
in `geometric_slerp()`, and update the documentation for
the interpolation parameter accordingly.

* The `Raises` section of the `geometric_slerp()` docstring
was also updated to remove a spurious claim about antipodes--we
now actually allow i.e., North and South pole points despite
the infinite number of paths between them on the sphere, and
enforce this behavior in our regression tests (a warning, but not error,
is issued). That is not directly related to the changes here, but
the documented innacurracy of behavior would otherwise make this
harder to reason about.

[ci skip] [skip ci]

Ockenfuss commented:
wow, thanks @tylerjereddy! Yes, I will have a look at the docs and tests in the next days.

tylerjereddy committed:
DOC: PR 23426 revisions

* Update the `geometric_slerp()` docstring to include an example
of the new extrapolation capability.

tylerjereddy commented on scipy/spatial/_geometric_slerp.py:
```diff
@@ -172,6 +175,26 @@ def geometric_slerp(
+    ...         result[...,1],
+    ...         result[...,2],
+    ...         c='k')
+    >>> plt.show()
```
I'll also let the CI flush through now, to see where we stand according to that.

tylerjereddy commented:
I don't think the free-threaded CI issue is related. `spatial.transform` has a Slerp function as well (in quaternion context). If worst comes to worst we could ask folks maintaining that for a review--there may be enough conceptual similarity that they could have an educated opinion. But this not urgent anyway, I'll move on to other stuff for now.

nickodell commented:
> I don't think the free-threaded CI issue is related.

I agree. I investigated the issue in more detail, and found that it was a bug in pytest-run-parallel which causes it to no longer be able to auto-detect that `test_pyyaml_not_found` is not thread safe. I've opened a new issue about it.

rgommers commented on scipy/spatial/tests/test_slerp.py:
```diff
@@ -416,3 +403,64 @@ def test_interpolation_param_ndim(self, t):
+        with pytest.warns(UserWarning, match='antipodes'):
+            actual_path = geometric_slerp(start=start,
+                                          end=end,
+                                          t=t)
```
fits on a single line

rgommers commented on scipy/spatial/tests/test_slerp.py:
```diff
@@ -416,3 +403,64 @@ def test_interpolation_param_ndim(self, t):
+    ([0, 0, 1], [0, 0, -1], -1, [0, 0, -1]),
+    ])
+    def test_extrapolation_antipodes(self, start, end, t, expected):
+        with pytest.warns(UserWarning, match='antipodes'):
```
This doesn't look quite right. If extrapolation is supported, it should be working without warnings being emitted. If this warning filter is for a subset of the test cases, can you please split the list of tests into two?

rgommers reviewed (commented):
Thanks @tylerjereddy, this overall looks good to me and the discussion about why to add this seems to have converged. A few comments.

@Ockenfuss did you want to look at this more?

Ockenfuss commented:
@rgommers I posted my version of the docs and examples [in this fork here](https://github.com/Ockenfuss/scipy/commit/23625936341ee6da0404514179445a5432890da9), but tylerjereddy's version also looks good to me. So there's nothing against a merge from my side!

tylerjereddy committed:
DOC: PR 23426 revisions

* Fix the line breaks in the new docstring example
based on reviewer feedback; also fix line breaks
in `test_extrapolation_antipodes()`.

* Indent the parametrization code/comments above
`test_extrapolation_basic()` by 4 spaces based
on reviewer feedback.

tylerjereddy commented:
@rgommers ok, I've revised to address all comments except for https://github.com/scipy/scipy/pull/23426#discussion_r2312305016. The antipodal warnings are intended and documented--we could remove them for both interpolated and extrapolated scenarios, but I'm not sure we should do that here.

The original discussion about this is at https://github.com/scipy/scipy/pull/10380#discussion_r336760595. It looks like I was pushing for an error for antipodes six years ago, but ended up with a warning. I still can't decide what is best--we return one of the infinite possible paths between poles in i.e. 3D, and the lack of a unique solution makes me think that a warning might still suit. I'm not super attached to that idea though, and I suppose removing a warning isn't a huge backcompat break, but this would also affect the original interpolation behavior added in 2020, so perhaps I could convince you to delay on that adjustment (both test cases in this new test are intentionally checking for this warning when extrapolating, for consistency with interpolating).

rgommers reviewed (approved):
Sure, I'm fine with your assessment that a warning for the antipodal usage is best - thanks for the context. 

All LGTM now, thanks @tylerjereddy and @Ockenfuss!
